File: _includes/sort-benchmark/algorithms/gravity.rs
fn gravity_sort(a: &mut [usize]) {
    if a.is_empty() {
        return;
    }

    let max = *a.iter().max().unwrap();
    if max == 0 {
        return;
    }

    let mut beads = vec![0usize; max];

    for &x in a.iter() {
        for bead in beads.iter_mut().take(x) {
            *bead += 1;
        }
    }

    let n = a.len();
    for i in (0..n).rev() {
        let mut sum = 0;
        for bead in beads.iter_mut() {
            if *bead == 0 {
                break;
            }
            sum += 1;
            *bead -= 1;
        }
        a[i] = sum;
    }
}

File: _includes/sort-benchmark/algorithms/gravity.swift
func gravity_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { gravity_sort($0) }
}

func gravity_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    if a.count == 0 {
        return
    }

    var max = a[0]
    for i in 1..<a.count {
        if a[i] > max {
            max = a[i]
        }
    }
    if max == 0 {
        return
    }

    var beads = [Int](repeating: 0, count: max)

    for i in 0..<a.count {
        let x = a[i]
        for j in 0..<x {
            beads[j] += 1
        }
    }

    let n = a.count
    for i in (0..<n).reversed() {
        var sum = 0
        for j in 0..<beads.count {
            if beads[j] == 0 {
                break
            }
            sum += 1
            beads[j] -= 1
        }
        a[i] = sum
    }
}
